[TFLite] Add INT8 and INT16x8 support to NON_MAX_SUPPRESSION_V4 / V5 operators

arovir01 opened the pull request:
This pull request adds a fully quantized INT8 and INT16x8 implementation for the TensorFlow Lite NON_MAX_SUPPRESSION_V4  and NON_MAX_SUPPRESSION_V5 operators.

arovir01 committed:
Add INT8 and INT16 support to NON_MAX_SUPPRESSION_V4/V5 operators

Change-Id: I67f70db0b16da1de31521e74990db3d1fdb8615e

arovir01 committed:
Add INT8 and INT16 support to NON_MAX_SUPPRESSION_V4/V5 operators

Change-Id: I67f70db0b16da1de31521e74990db3d1fdb8615e

abattery commented:
Could you share why these ops are needed for your case to understand your use cases? e.g., a certain model should have int16x8. We are trying to keep the TFLite binary slim if possible if there are no valid cases.

arovir01 reviewed (commented):
> Could you share why these ops are needed for your case to understand the contexts around this PR? e.g., a certain model should have int16x8 op support like this one.
> 
> We are trying to keep the TFLite binary slim if possible if there are no valid use cases.

Hi,

This is part of our effort to support variants of the EfficientDet object detection model on integer-only hardware. These NMS operators are also at the core of our investigation around supporting MobileNet models that use the custom TFLite_Detection_PostProcess operator.

abattery commented:
Thanks for sharing the contexts! @jianlijianli @thaink 
Could you review this PR?

arovir01 committed:
Improve overflow protection in quantized NON_MAX_SUPPRESSION implementation

arovir01 committed:
Improve overflow protection in quantized NON_MAX_SUPPRESSION implementation

arovir01 committed:
Merge branch 'master' into toupstream/quantized_non_max_suppression

gbaned commented:
@arovir01 Can you please resolve conflicts? Thanks!

arovir01 committed:
Merge branch 'master' into toupstream/quantized_non_max_suppression

gbaned commented:
@jianlijianli Can you please review this PR ? Thanks!

arovir01 commented:
I suspect the versioning needs to be updated from 2.7 to at least 2.8, as 2.7 has already since been released.

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Update version number on NON_MAX_SUPPRESSION_V4/V5

arovir01 committed:
Update version number on NON_MAX_SUPPRESSION_V4/V5

arovir01 commented:
> I suspect the versioning needs to be updated from 2.7 to at least 2.8, as 2.7 has already since been released.

Done.

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Merge branch 'master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Add QuantizableResult trait to NON_MAX_SUPPRESSION_V5

arovir01 committed:
Add QuantizableResult trait to NON_MAX_SUPPRESSION_V5

Tessil committed:
Quantize the inputs of the INT16 NON_MAX_SUPPRESSION_V5 op as activations

Tessil committed:
Quantize the inputs of the INT16 NON_MAX_SUPPRESSION_V5 op as activations

Tessil committed:
Remove unused boxes_zero_point variable in NonMaxSuppressionParams

Change-Id: I291729e82a2ab775cef1892e1cedab6d825429ee

Tessil committed:
Remove unused boxes_zero_point variable in NonMaxSuppressionParams

Change-Id: I291729e82a2ab775cef1892e1cedab6d825429ee

Tessil commented:
For information here are some quick results of the Yolo v3 416x416 model using the quantized NMS op of this PR with the COCO 2014 dataset. The whole model is quantized except the EXP operator (for which a TABLE op could be used). 

We used 1000 images from the training set for calibration and 2000 images of the validation set for the tests as we don't have the ground truth values for coco-test (only 2000 images were used for a quick test as some of the int16 kernels are not yet well optimized and the whole set would take a long time). Score threshold of 0.3, max boxes of 100 and an IoU of 0.5 is used for the mAP score.

| Type | mAP  |
| ----- |-------------:|
| float32 | 0.566 |
| int8   | 0.523 |
| int16 |  0.565 |

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 committed:
Merge branch 'toupstream/quantized_non_max_suppression' of github.com:arovir01/tensorflow into toupstream/quantized_non_max_suppression

arovir01 committed:
Merge branch 'tensorflow:master' into toupstream/quantized_non_max_suppression

arovir01 commented:
Done.

arovir01 commented:
Since I have moved on to another team & project, I have also added @Tessil as a collaborator. This should allow him to fix merge conflicts and push commits whenever required.

gbaned commented:
@Tessil  Can you please resolve conflicts? Thank you!

Tessil committed:
Fix errors made during merge-conflict resolution

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/quantized_non_max_suppression

gbaned commented:
Hi @jianlijianli /@thaink Can you please review this PR ? Thank you!

gbaned commented:
@Tessil Can you please resolve conflicts? Thank you!

gbaned commented:
@arovir01 Can you please resolve conflicts? Thank you!

arovir01 commented:
> @arovir01 Can you please resolve conflicts? Thank you!

I am no longer involved with this work, as mentioned [above](https://github.com/tensorflow/tensorflow/pull/51785#issuecomment-1084622764). Please contact @Tessil for conflict resolution or other requests.

Tessil committed:
Merge branch 'master' into toupstream/quantized_non_max_suppression

Tessil committed:
Replace selected_scores() by getSelectedScores(), prefixed getters must now be used

Tessil committed:
Add support for passing extra parameters to the LUTPopulate transform function pointer

Tessil committed:
Adapt to LUT generation refactoring

Tessil commented:
@gbaned I resolved the conflicts.

Note that the code from https://github.com/tensorflow/tensorflow/pull/58518 has been imported into this PR to make it work properly. https://github.com/tensorflow/tensorflow/pull/58518 should be merged before this one.

gbaned commented:
Hi @arovir01 Can you please resolve conflicts? Thank you!

arovir01 commented:
> Hi @arovir01 Can you please resolve conflicts? Thank you!

Forwarded to @Tessil.

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/quantized_non_max_suppression

Tessil committed:
Merge remote-tracking branch 'upstream/master' into toupstream/quantized_non_max_suppression

gbaned commented:
Hi @arovir01 Can you please resolve conflicts? Thank you!

arovir01 committed:
Merge branch 'master' into toupstream/quantized_non_max_suppression

psunn committed:
Update NMS op version with current runtime version

mihaimaruseac commented:
At some point we might want to close this PR as abandoned given it's from 2021 and keeps failing to land

arovir01 commented:
> At some point we might want to close this PR as abandoned given it's from 2021 and keeps failing to land

I believe some of my colleagues have resurrected this change and might still be aiming to merge it.

psunn commented:
@Tessil FYI

> At some point we might want to close this PR as abandoned given it's from 2021 and keeps failing to land
@mihaimaruseac Hey Mihai, We've been trying to merge this PR for quite a while now. Could you please elaborate on what you mean by "keeps failing to land"? This way, we can try to figure out the blocker and see how we can help on our side.

mihaimaruseac commented:
The PR doesn't even run internal CI tests as a gating presubmit is failing. No one looked at the error from there to shepherd it forward

mihaimaruseac committed:
Update tensorflow/lite/kernels/non_max_suppression.cc

mihaimaruseac commented on tensorflow/lite/kernels/internal/reference/non_max_suppression.h:
```diff
@@ -18,111 +18,241 @@ limitations under the License.
 
 namespace tflite {
 namespace reference_ops {
+namespace {
```
Please don't use internal namespaces in header files.

https://google.github.io/styleguide/cppguide.html#Internal_Linkage

Should this and the implementations be moved to a `.cc` file instead?

Tessil replied:
I renamed the namespace to `namespace detail` as done in other parts of TFLite.

If preferred, I can check to put the implementation in a `.cc` file and instantiate the template there.

mihaimaruseac commented:
I just ran all internal presubmits that could run, there's only one lint error mentioned above. You will need to fix that and then we can run more.

Tessil committed:
Rename anonymous namespace to detail

mihaimaruseac commented:
Let's try this way for now

mihaimaruseac commented:
```
In file included from [tensorflow/lite/kernels/non_max_suppression.cc:15]:
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:244]:7: error: use of undeclared identifier 'GetCandidatePriorityQueue'; did you mean 'detail::GetCandidatePriorityQueue'?
      GetCandidatePriorityQueue(scores, num_boxes, score_threshold);
      ^~~~~~~~~~~~~~~~~~~~~~~~~
      detail::GetCandidatePriorityQueue
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:171]:6: note: 'detail::GetCandidatePriorityQueue' declared here
auto GetCandidatePriorityQueue(const float* scores, const int num_boxes,
     ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:255]:5: error: no template named 'Candidate'; did you mean 'detail::Candidate'?
    Candidate<float> next_candidate = candidate_priority_queue.top();
    ^~~~~~~~~
    detail::Candidate
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:161]:8: note: 'detail::Candidate' declared here
struct Candidate {
       ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:271]:11: error: use of undeclared identifier 'ComputeIou'; did you mean 'detail::ComputeIou'?
          ComputeIou(boxes, next_candidate.index, selected_indices[j]);
          ^~~~~~~~~~
          detail::ComputeIou
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:98]:14: note: 'detail::ComputeIou' declared here
inline float ComputeIou(const float* boxes, int i, int j) {
             ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:356]:7: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
      Rescale<T>(iou_threshold, params.iou_threshold_zero_point,
      ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:363]:7: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
      Rescale<T>(score_threshold, params.score_threshold_zero_point,
      ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:381]:5: error: no template named 'Candidate'; did you mean 'detail::Candidate'?
    Candidate<T> next_candidate = candidate_priority_queue.top();
    ^~~~~~~~~
    detail::Candidate
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:161]:8: note: 'detail::Candidate' declared here
struct Candidate {
       ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:417]:32: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++20 extension [-Wc++20-extensions]
        next_candidate.score = Truncate<T>(
                               ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:356]:7: error: use of undeclared identifier 'Rescale'
      Rescale<T>(iou_threshold, params.iou_threshold_zero_point,
      ^
[tensorflow/lite/kernels/non_max_suppression.cc:464]:20: note: in instantiation of function template specialization 'tflite::reference_ops::NonMaxSuppression<signed char>' requested here
    reference_ops::NonMaxSuppression(
                   ^
[tensorflow/lite/kernels/non_max_suppression.cc:515]:14: note: in instantiation of function template specialization 'tflite::ops::builtin::non_max_suppression::Eval<signed char>' requested here
      return Eval<int8_t>(context, node);
             ^
In file included from [tensorflow/lite/kernels/non_max_suppression.cc:15]:
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:363]:7: error: use of undeclared identifier 'Rescale'
      Rescale<T>(score_threshold, params.score_threshold_zero_point,
      ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:390]:11: error: use of undeclared identifier 'ComputeIou'
          ComputeIou(boxes, next_candidate.index, selected_indices[j], params);
          ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:417]:32: error: use of undeclared identifier 'Truncate'
        next_candidate.score = Truncate<T>(
                               ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:356]:7: error: use of undeclared identifier 'Rescale'
      Rescale<T>(iou_threshold, params.iou_threshold_zero_point,
      ^
[tensorflow/lite/kernels/non_max_suppression.cc:464]:20: note: in instantiation of function template specialization 'tflite::reference_ops::NonMaxSuppression<short>' requested here
    reference_ops::NonMaxSuppression(
                   ^
[tensorflow/lite/kernels/non_max_suppression.cc:517]:14: note: in instantiation of function template specialization 'tflite::ops::builtin::non_max_suppression::Eval<short>' requested here
      return Eval<int16_t>(context, node);
             ^
In file included from [tensorflow/lite/kernels/non_max_suppression.cc:15]:
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:363]:7: error: use of undeclared identifier 'Rescale'
      Rescale<T>(score_threshold, params.score_threshold_zero_point,
      ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:390]:11: error: use of undeclared identifier 'ComputeIou'
          ComputeIou(boxes, next_candidate.index, selected_indices[j], params);
          ^
[./tensorflow/lite/kernels/internal/reference/non_max_suppression.h:417]:32: error: use of undeclared identifier 'Truncate'
        next_candidate.score = Truncate<T>(
```

Tessil committed:
Adapt calls to namespace rename

mihaimaruseac commented:
A warning but fails on builds with `-Werror`:

```
'tflite::(anonymous namespace)::BaseNMSOp' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
   58 | class BaseNMSOp : public SingleOpModel {
      |       ^
```

mihaimaruseac commented on tensorflow/lite/kernels/non_max_suppression_test.cc:
```diff
@@ -26,20 +27,106 @@ namespace {
+    std::minmax_element(k_input_boxes_data.begin(), k_input_boxes_data.end());
+
 class BaseNMSOp : public SingleOpModel {
  public:
```
```suggestion
 public:
   virtual ~BaseNMSOp() {}
 
```

I think this should fix the current issue

mihaimaruseac committed:
Update tensorflow/lite/kernels/non_max_suppression_test.cc

mihaimaruseac commented:
The OSS tests seem to be passing, however, we're seeing duplicate symbols during linking on internal builds that are able to detect this (OSS build is not able to, opening up to ODR violations)

```
ld.lld: error: duplicate symbol: tflite::reference_ops::detail::GetCandidatePriorityQueue(float const*, int, float)
>>> defined at non_max_suppression_test.cc
>>>            bazel-out/arm-opt/bin/tensorflow/lite/kernels/internal/_objs/non_max_suppression_test/non_max_suppression_test.o:(tflite::reference_ops::detail::GetCandidatePriorityQueue(float const*, int, float))
>>> defined at non_max_suppression.cc
>>>            non_max_suppression.o:(.text._ZN6tflite13reference_ops6detail25GetCandidatePriorityQueueEPKfif+0x0) in archive bazel-out/arm-opt/bin/tensorflow/lite/kernels/libbuiltin_op_kernels.a
```

Tessil committed:
Fix duplicate symbols in NMS helpers functions

mihaimaruseac commented:
Closing this PR as it has been failing to build internally for nearly 3 years
